typing/env.ml

```diff
@@ -1017,18 +1017,25 @@ let add_required_global id =
   && not (List.exists (Ident.same id) !required_globals)
   then required_globals := id :: !required_globals
 
-let rec normalize_path lax env path =
-  let path =
-    match path with
-      Pdot(p, s, pos) ->
-        Pdot(normalize_path lax env p, s, pos)
-    | Papply(p1, p2) ->
-        Papply(normalize_path lax env p1, normalize_path true env p2)
-    | _ -> path
-  in
+let rec normalize_module_path lax env = function
+  | Pident id as path when lax && Ident.persistent id ->
+      path (* fast path (avoids lookup) *)
+  | Pdot (p, s, pos) as path ->
+      let p' = normalize_module_path lax env p in
+      if p == p' then normalize_module_path0 lax env path
+      else normalize_module_path0 lax env (Pdot(p', s, pos))
+  | Papply (p1, p2) as path ->
+      let p1' = normalize_module_path lax env p1 in
+      let p2' = normalize_module_path true env p2 in
+      if p1 == p1' && p2 == p2' then normalize_module_path0 lax env path
+      else normalize_module_path0 lax env (Papply(p1', p2'))
+  | Pident _ as path ->
+      normalize_module_path0 lax env path
+
```

Comment:
What about calling this function `expand_module_path` rather `normalize_module_path0` ?